### ISSUE TITLE
The model can not converge.

### ISSUE THREAD

**Issue by vzxxbacq** [omoindrot/tensorflow-triplet-loss#15]
Hello @omoindrot , thanks for your great work. I build a Resnet to do speaker recognition job, but when I use triplet loss to train my model the loss is 0.5 and didn't decrease. Am I use triplet loss in wrong way?

```python
    def _inference(self, inp):
        for i in range(self.n_blocks):
            if i > 0:
                inp_channel = inp.get_shape().as_list()[-1]
                inp = self._conv2d(name='conv5%d'%i, shape=[5, 5, inp_channel, self.out_channel[i]],
                                   strides=[1, 2, 2, 1], x=inp, padding='SAME')
                inp = self._residual_block(inp, self.out_channel[i], "residual_block_%d" % i,
                                           is_first_layer=False)

            else:
                inp_channel = inp.get_shape().as_list()[-1]
                inp = self._conv2d(name='conv5%d' % i, shape=[5, 5, inp_channel, self.out_channel[i]],
                                   strides=[1, 2, 2, 1], x=inp, padding='SAME')
                inp = self._residual_block(inp, self.out_channel[i], "residual_block_%d" % i,
                                           is_first_layer=True)
            print(inp.get_shape().as_list())

        inp = tf.reduce_mean(inp, axis=1)

        inp = tf.reshape(inp, [-1, inp.get_shape().as_list()[1]*inp.get_shape().as_list()[2]])
        
        weight_affine = self._new_variable("affine_weight", [inp.get_shape().as_list()[-1], 512],
                                           weight_type="FC")
        bias_affine = self._new_variable("affine_bias", [512], "FC")
        inp = tf.nn.relu(tf.matmul(inp, weight_affine) + bias_affine)
        print(inp.get_shape().as_list())
        output = tf.nn.l2_normalize(inp)
        return output
```
```python
    def _triplet_loss(self, inp, targets):
        if targets.get_shape().as_list()[-1] != 1:
            targets = tf.argmax(targets, axis=1)
        loss = triplet_loss.batch_hard_triplet_loss(targets, inp, 0.5)
        return loss
```
![image](https://user-images.githubusercontent.com/29952934/46590259-75a67e80-cae4-11e8-9205-b2e024c2f4b5.png)

**Comment by omoindrot** [comment]
See issue #6, if the loss is exactly `0.5` it means the embeddings have collapse into one point.

The first thing to try is to decrease the learning rate, it worked for other people before.

An additional note: the loss will not decrease during training because it's the average over the positive triplets. So the loss will just oscillate during training. However the fraction of positive triplets should decrease.

**Comment by vzxxbacq** [comment]
thanks for your advice.
